packages/core/src/common/refs.ts

```diff
@@ -38,3 +38,32 @@ export function getRef<T = HTMLElement>(ref: T | IRefObject<T> | null) {
 
     return (ref as IRefObject<T>).current ?? (ref as T);
 }
+
+export function setRef<T extends HTMLElement>(refTarget: IRef<T>, ref: T | null) {
+    if (isRefObject<T>(refTarget)) {
+        refTarget.current = ref;
+    }
+    if (isRefCallback(refTarget)) {
+        refTarget(ref);
+    }
+}
+
+export function refHandler<T extends HTMLElement>(
+    refProp: IRef<T> | undefined | null,
+    refTargetParent: Record<string, any>,
+    refTargetKey: string,
+) {
+    if (!(refTargetKey in refTargetParent)) {
+        throw new Error(`${refTargetKey} does not exists in ${Object.keys(refTargetParent)}`);
```

Comment:
I really don't like that we've moved an error from compilation time that is guarded by type checking to one that is thrown at runtime. That goes against the goals of using TypeScript. I would be open to another solution which has a small drawback of requiring all ref fields to be public:

```ts
export function refHandler<T extends HTMLElement, K extends string>(
    refProp: IRef<T> | undefined | null,
    refTargetParent: { [k in K]: T | IRefObject<T> | null },
    refTargetKey: K,
) {
    if (isRefObject<T>(refProp)) {
        refTargetParent[refTargetKey] = refProp;
        return refProp;
    }
    return (ref: T | null) => {
        refTargetParent[refTargetKey] = ref;
        if (isRefCallback(refProp)) {
            refProp(ref);
        }
    };
}
```

example usage in Checkbox:

```ts
    public input: HTMLInputElement | null = null;
    private handleInputRef: IRef<HTMLInputElement> = refHandler(this.props.inputRef, this, "input");
```

note the type annotation is required on `handleInputRef` because otherwise the inferred type has a cyclical reference which cannot be computed.

Reply:
Yes, I originally wanted to do refTargetKey as ```keyof T``` but Typescript doesn't expose private properties on ```keyof```. Didn't made ref public as thought original implementation would be preferred (since users are not modifying the classes and the tests would fail if an invalid key was given), however I don't mind changing refs to public.

Reply:
The developer experience concern (it's 2021, we should not allow typos in code uncaught by the compiler lol) outweighs the private encapsulation benefit (getting a reference to the class instance is nonstandard usage anyway). Go ahead with changing them to public